### ISSUE TITLE
Generating multiple sizes only uses the last size for the entire set

### ISSUE THREAD

**Issue by qelix** [JonasKruckenberg/imagetools#140]
Hey, another one from me ;)

I discovered a serious bug that actually completely breaks the source set generation. If you generate multiple images by chaining multiple widths together (e.g. `import srcsetAvif from '../example.jpg?w=500;700;900;1200&avif&srcset'`, as per your example), it actually only applies the last width on every image in the whole set!

To reproduce: 

1. Clone this repo and `cd` into `examples/vite-simple`
2. Edit `main.js` and add ridiculously low widths as the last params: 

```js
// ...
import srcsetAvif from '../example.jpg?w=500;700;900;1200;20&avif&srcset'
import srcsetWebp from '../example.jpg?w=500;700;900;1200;20&webp&srcset'
// ...
```
3. `yarn && yarn dev`
4. Open the page in your browser, open the devtools and reload
5. In your network tab, you'll now see the exact file that was loaded - for me it had an ID of `e78e253f7e1877238273f8472026215d31af4221` (you guessed it: it's only 20px wide 😉)
6. When inspecting srcset of the the image, however, you'll notice that the loaded ID was supposed to be 1200px wide (depending on your viewport, of couse): `/@imagetools/e78e253f7e1877238273f8472026215d31af4221 1200w,`. The one with 20px has an entirely different ID.

So, even though the 20px wide image was not selected by the browser, it still only received a 20px wide one. It looks like it's always the last argument that decides the size of all the images in the set. All of the metadata is still generated according to the parameters. Each URL is just pointing to an image the size of the last parameter.

**Comment by JonasKruckenberg** [comment]
This is really bad 😟
This only appears during development right? So when you do yarn build all the right images get generated?

**Comment by qelix** [comment]
Oh, I haven't even thought about testing that. And yeah, you're right, it works in production, so it isn't that terrible after all.

**Comment by bronze** [comment]
>  This only appears during development right? So when you do yarn build all the right images get generated?

oh god dammit... I've been cracking my head trying to figure this one out...
yeah.... can confirm, at least on my testing, the switch happens alright once deployed, just during dev that it only shows the last size

**Comment by jugglingcats** [comment]
@JonasKruckenberg I just hit this one too. The problem seems to be in the vite plugin and the middleware added in `configureServer`. To debug I looked for use of `generatedImages` and added the following:

```
// before generatedImage.set
console.log("set image", id, imagetoolsCore.getMetadata(image, 'width'))

// after generatedImage.get (in middleware)
console.log("get image", id, imagetoolsCore.getMetadata(image, 'width'))
```

In the logs...
```
set image 6269821afe096f8aea2f2ba455cddea6fa0b734a 500
set image cb68b63cfb6eadc77da1ba1b4f5866595fe545e2 700
get image 6269821afe096f8aea2f2ba455cddea6fa0b734a 700 <--- ooops!
get image cb68b63cfb6eadc77da1ba1b4f5866595fe545e2 700
```

Something is being reused somewhere...? I then modified this line:
```
//             generatedImages.set(id, image); // from
                generatedImages.set(id, image.clone()); // to
```

New log output:
```
set image 6269821afe096f8aea2f2ba455cddea6fa0b734a 500
set image cb68b63cfb6eadc77da1ba1b4f5866595fe545e2 700
get image 6269821afe096f8aea2f2ba455cddea6fa0b734a undefined
get image cb68b63cfb6eadc77da1ba1b4f5866595fe545e2 undefined
```
... and images in the site are the correct size 🎉.

**Comment by jugglingcats** [comment]
I guess it is because of this line:
```
            const img = imagetoolsCore.loadImage(decodeURIComponent(srcURL.pathname));
```
... this `img` is passed to `imagetoolsCore.applyTransforms` and returned, so same image is put in the `Set`.

**Comment by JonasKruckenberg** [comment]
Good catch! The sharp instance is mutable indeed and so will be changed by subsequent iterations, cloning should solve this!

**Comment by suamai** [comment]
Hey, this issue seems to have been fixed by the PR #218 , but was rolled back by this commit f6cec968782c13b1b9d35648d11aa09822ce2d05 .

Is there a reason for that?

**Comment by JonasKruckenberg** [comment]
Oups, must have been lost in the merges somehow!

**Comment by JonasKruckenberg** [comment]
Fixed and published please upgrade to `vite-imagetools@4.0.1`
